controllers/clustermanagementaddon_controller.go

```diff
@@ -0,0 +1,590 @@
+	"open-cluster-management.io/cluster-proxy/pkg/addon/operator/eventhandler"
+
+	"github.com/openshift/library-go/pkg/crypto"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	appsv1client "k8s.io/client-go/kubernetes/typed/apps/v1"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var _ reconcile.Reconciler = &ClusterManagementAddonReconciler{}
+
+var log = ctrl.Log.WithName("ClusterManagementAddonReconciler")
+
+type ClusterManagementAddonReconciler struct {
+	client.Client
+	SelfSigner       selfsigned.SelfSigner
+	CAPair           *crypto.CA
+	SecretLister     corev1listers.SecretLister
+	SecretGetter     corev1client.SecretsGetter
+	DeploymentGetter appsv1client.DeploymentsGetter
+	ServiceGetter    corev1client.ServicesGetter
+	EventRecorder    events.Recorder
+
+	proxyConfigClient  proxyclient.Interface
+	proxyConfigLister  proxylister.ManagedProxyConfigurationLister
+	addonLister        addonlister.ManagedClusterAddOnNamespaceLister
+	clusterAddonLister addonlister.ClusterManagementAddOnLister
+}
+
+func (c *ClusterManagementAddonReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&addonv1alpha1.ClusterManagementAddOn{}).
+		Watches(
+			&source.Kind{
+				Type: &proxyv1alpha1.ManagedProxyConfiguration{},
+			},
+			&eventhandler.ManagedProxyConfigurationHandler{
+				Client: mgr.GetClient(),
+			}).
+		Complete(c)
+}
+
+func (c *ClusterManagementAddonReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log.Info("Start reconcile", "name", request.Name)
+
+	// get the latest cluster-addon
+	addon := &addonv1alpha1.ClusterManagementAddOn{}
+	if err := c.Client.Get(ctx, request.NamespacedName, addon); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("Cannot find cluster-addon", "name", request.Name)
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	if len(addon.Spec.AddOnConfiguration.CRName) == 0 {
+		log.Info("Skipping cluster-addon, no config coordinate", "name", request.Name)
+		return reconcile.Result{}, nil
+	}
+
+	// get the related proxy configuration
+	config := &proxyv1alpha1.ManagedProxyConfiguration{}
+	if err := c.Client.Get(ctx, types.NamespacedName{
+		Name: addon.Spec.AddOnConfiguration.CRName,
+	}, config); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("Cannot find proxy-configuration", "name", addon.Spec.AddOnConfiguration.CRName)
```

Comment:
We may conside the deletaion case, if there is no config, we may need to cleean up the deploy

Reply:
the deletion is working by linking the proxy config as owner to those provisioned resources i.e. these resources will be cleaned upon proxy config's deletion. wdyt?